src/types/list.md

```diff
@@ -1 +1,66 @@
 # List
+
+Lists are the SML type of ordered collections of objects. Notably, you can create lists of any kind of object, so `int list`, `string list`, and `bool list` are all valid types. `list` on its own is a type constructor (meaning that it makes new types out of old types), so it is not a valid type by itself, however. Lists are not, however, the same as arrays - they do not have constant-time access to any given index of the list. Instead, you only have access to the elements located at the very beginning of the list, the _head_. Additionally, lists are more _restricted_ than data structures in some other languages - a given list has a fixed type for its elements. All of the elements in a list must be of the same type. For the purposes of this document, we will discuss only int lists.
+
+## Values
+
+We write lists as a sequence of integers, separated by commas, all enclosed with two square brackets. So then we have valid int lists as `[1, 2, 3]`, `[1, 5, 1, 5, 0]`, and `[]`, with the latter representing the empty list. We also refer to the empty list as _nil_ (in addition, you can type `nil` instead of `[]` in code). 
+
+The other essential component to lists is what is known as the `::` operator, referred to as "cons".
+```sml
+(op ::) : int * int list -> int list
+```
+
+Cons takes in an int and an int list, and prepends the int to the front of the given int list. For instance, we have that `1 :: [2, 3]` steps to `[1, 2, 3]`. Additionally, cons is _right-associative_. This means that in a continuous stream of applications of cons, they are evaluated from _right-to-left_. This means that `1 :: 2 :: 3 :: []` is implicitly denoting `1 :: (2 :: (3 :: []))`, as the calls to cons associate to the right. So, similarly to before, a `1 :: [2, 3]`, `[1, 2, 3]`, and `1 :: 2 :: 3 :: []` are exactly equivalent, and denote the same list.
+
+Cons is also a _constructor_, meaning that it can be used to pattern match and deconstruct lists. As such, we can write basic functions to compute the length of a list as follows:
+```sml
+(* length : int list -> int *)
+(* REQUIRES: true *)
+(* ENSURES: length L ==> the length of L *)
+fun length ([] : int list) : int = 0
+  | length (x::xs : int list) : int = 1 + length xs
+```
+Given a non-empty list, this function simply binds the first element of the list to the identifier `x`, discards it, and then recursively calls `length` to find the length of the remaining list `xs`, adding 1 to the result.
+
+The definition of an int list thus corresponds to:
+```sml
+datatype int list = [] | :: of int * int list
```

Comment:
Other than gripes with talking only about ints, this one is actually invalid syntax; while I'm all for lying to beginners, I think this is a bit of a line being crossed.

Reply:
That's why I wrote "corresponds to" - though I could add a disclaimer saying that this would not actually work in the SML/NJ REPL, afterwards. Thoughts?